pkg/controller/pipeline.go

```diff
@@ -0,0 +1,523 @@
+)
+
+const (
+	LabelOwnedByCapsule = "rig.dev/owned-by-capsule"
+)
+
+func GetNew[T client.Object](req Request, key ObjectKey) T {
+	var t T
+	o := req.GetNew(key)
+	if c, ok := o.(T); ok {
+		t = c
+	}
+	return t
+}
+
+func GetCurrent[T client.Object](req Request, key ObjectKey) T {
+	var t T
+	o := req.GetCurrent(key)
+	if c, ok := o.(T); ok {
+		t = c
+	}
+	return t
+}
+
+func Get[T interface {
+	client.Object
+	comparable
+}](req Request, key ObjectKey,
+) T {
+	var t T
+	if o := GetNew[T](req, key); o != t {
+		return o
+	}
+	if o := GetCurrent[T](req, key); o != t {
+		return o
+	}
+	return t
+}
+
+type Object struct {
+	Current client.Object
+	New     client.Object
+}
+
+type ObjectKey struct {
+	client.ObjectKey
+	schema.GroupVersionKind
+}
+
+func (ok ObjectKey) String() string {
+	return fmt.Sprintf("%s/%s", ok.GroupVersionKind.String(), ok.ObjectKey.String())
+}
+
+func (ok ObjectKey) MarshalLog() interface{} {
+	return struct {
+		Group     string `json:"group,omitempty"`
+		Version   string `json:"version,omitempty"`
+		Kind      string `json:"kind"`
+		Name      string `json:"name"`
+		Namespace string `json:"namespace,omitempty"`
+	}{
+		Name:      ok.Name,
+		Namespace: ok.Namespace,
+		Group:     ok.Group,
+		Kind:      ok.Kind,
+		Version:   ok.Version,
+	}
+}
+
+func Equal(a ObjectKey, b ObjectKey) bool {
+	return a == b
+}
+
+type Request interface {
+	Config() *configv1alpha1.OperatorConfig
+	Scheme() *runtime.Scheme
+	Client() client.Client
+	Capsule() *v1alpha2.Capsule
+	GetCurrent(key ObjectKey) client.Object
+	GetNew(key ObjectKey) client.Object
+	Set(key ObjectKey, obj client.Object)
+	NamedObjectKey(name string, gvk schema.GroupVersionKind) ObjectKey
+	ObjectKey(gvk schema.GroupVersionKind) ObjectKey
+	MarkUsedResource(res v1alpha2.UsedResource)
+}
+
+type Step interface {
+	Apply(ctx context.Context, req Request) error
+}
+
+type Pipeline struct {
+	client         client.Client
+	config         *configv1alpha1.OperatorConfig
+	scheme         *runtime.Scheme
+	logger         logr.Logger
+	capsule        *v1alpha2.Capsule
+	currentObjects map[ObjectKey]client.Object
+	objects        map[ObjectKey]*Object
+	steps          []Step
+	generation     int64
+	usedResources  []v1alpha2.UsedResource
+}
+
+func NewPipeline(
+	cc client.Client,
+	config *configv1alpha1.OperatorConfig,
+	capsule *v1alpha2.Capsule,
+	scheme *runtime.Scheme,
+	logger logr.Logger,
+) *Pipeline {
+	p := &Pipeline{
+		client: cc,
+		config: config,
+		scheme: scheme,
+		logger: logger.WithValues(
+			"capsule", capsule.Name,
+		),
+		capsule:        capsule,
+		currentObjects: map[ObjectKey]client.Object{},
+	}
+	if capsule.Status != nil {
+		p.generation = capsule.Status.ObservedGeneration
+	}
+	return p
+}
+
+func (p *Pipeline) Config() *configv1alpha1.OperatorConfig {
+	return p.config
+}
+
+func (p *Pipeline) Scheme() *runtime.Scheme {
+	return p.scheme
+}
+
+func (p *Pipeline) Capsule() *v1alpha2.Capsule {
+	return p.capsule
```

Comment:
Should we return deepcopies here (and for `Config` to ensure plugins aren't naughty and modifies stuff?

Reply:
Sure, will do